File: go/memory_store.go
package agentledger

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"iter"
	"sort"
	"sync"
	"time"
)

var (
	ErrLaneConflict         = errors.New("lane sequence conflict")
	ErrIdempotencyViolation = errors.New("append id reused with different content")
	ErrDuplicateEvent       = errors.New("duplicate event id")
	ErrEntityConflict       = errors.New("entity conflict")
	ErrEntityNotFound       = errors.New("entity not found")
	ErrSubjectMismatch      = errors.New("event subject does not belong to lane")
)

type MemoryEventStore struct {
	mu             sync.Mutex
	actors         map[string]Actor
	lanes          map[string]Lane
	laneNames      map[string]string
	turns          map[string]Turn
	actions        map[string]Action
	attempts       map[string]Attempt
	attemptNumbers map[string]struct{}
	events         map[string]StoredEvent
	laneEvents     map[string][]StoredEvent
	appends        map[string]AppendReceipt
}

func NewMemoryEventStore() *MemoryEventStore {
	return &MemoryEventStore{
		actors: make(map[string]Actor), lanes: make(map[string]Lane), laneNames: make(map[string]string),
		turns: make(map[string]Turn), actions: make(map[string]Action),
		attempts: make(map[string]Attempt), attemptNumbers: make(map[string]struct{}),
		events: make(map[string]StoredEvent), laneEvents: make(map[string][]StoredEvent),
		appends: make(map[string]AppendReceipt),
	}
}

func (s *MemoryEventStore) CreateActor(ctx context.Context, actor Actor) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.actors[actor.ID]; ok {
		return fmt.Errorf("%w: actor %s", ErrEntityConflict, actor.ID)
	}
	s.actors[actor.ID] = actor
	return nil
}

func (s *MemoryEventStore) GetActor(ctx context.Context, id string) (Actor, bool, error) {
	if err := ctx.Err(); err != nil {
		return Actor{}, false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	value, ok := s.actors[id]
	return value, ok, nil
}

func (s *MemoryEventStore) CreateLane(ctx context.Context, lane Lane) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	if lane.LastSeq != 0 {
		return errors.New("new lane must have last_seq 0")
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	nameKey := laneNameKey(lane.SessionID, lane.RunID, lane.Name)
	if _, ok := s.lanes[lane.ID]; ok {
		return fmt.Errorf("%w: lane %s", ErrEntityConflict, lane.ID)
	}
	if _, ok := s.laneNames[nameKey]; ok {
		return fmt.Errorf("%w: lane %s", ErrEntityConflict, lane.ID)
	}
	if lane.ParentLaneID != "" {
		parent, ok := s.lanes[lane.ParentLaneID]
		if !ok {
			return fmt.Errorf("%w: parent lane %s", ErrEntityNotFound, lane.ParentLaneID)
		}
		if parent.SessionID != lane.SessionID {
			return errors.New("parent lane must belong to same session")
		}
	}
	s.lanes[lane.ID] = lane
	s.laneNames[nameKey] = lane.ID
	return nil
}

func (s *MemoryEventStore) GetLane(ctx context.Context, id string) (Lane, bool, error) {
	if err := ctx.Err(); err != nil {
		return Lane{}, false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	value, ok := s.lanes[id]
	return value, ok, nil
}

func (s *MemoryEventStore) FindLane(ctx context.Context, sessionID, runID, name string) (Lane, bool, error) {
	if err := ctx.Err(); err != nil {
		return Lane{}, false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	id, ok := s.laneNames[laneNameKey(sessionID, runID, name)]
	if !ok {
		return Lane{}, false, nil
	}
	return s.lanes[id], true, nil
}

func (s *MemoryEventStore) CreateTurn(ctx context.Context, turn Turn) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.turns[turn.ID]; ok {
		return fmt.Errorf("%w: turn %s", ErrEntityConflict, turn.ID)
	}
	if _, ok := s.lanes[turn.LaneID]; !ok {
		return fmt.Errorf("%w: lane %s", ErrEntityNotFound, turn.LaneID)
	}
	s.turns[turn.ID] = turn
	return nil
}

func (s *MemoryEventStore) GetTurn(ctx context.Context, id string) (Turn, bool, error) {
	if err := ctx.Err(); err != nil {
		return Turn{}, false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	value, ok := s.turns[id]
	return value, ok, nil
}

func (s *MemoryEventStore) CreateAction(ctx context.Context, action Action) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.actions[action.ID]; ok {
		return fmt.Errorf("%w: action %s", ErrEntityConflict, action.ID)
	}
	if _, ok := s.turns[action.TurnID]; !ok {
		return fmt.Errorf("%w: turn %s", ErrEntityNotFound, action.TurnID)
	}
	if action.ParentActionID != "" {
		parent, ok := s.actions[action.ParentActionID]
		if !ok {
			return fmt.Errorf("%w: parent action %s", ErrEntityNotFound, action.ParentActionID)
		}
		if parent.TurnID != action.TurnID {
			return errors.New("parent action must belong to same turn")
		}
	}
	s.actions[action.ID] = action
	return nil
}

func (s *MemoryEventStore) GetAction(ctx context.Context, id string) (Action, bool, error) {
	if err := ctx.Err(); err != nil {
		return Action{}, false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	value, ok := s.actions[id]
	return value, ok, nil
}

func (s *MemoryEventStore) CreateAttempt(ctx context.Context, attempt Attempt) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	if attempt.AttemptNo < 1 {
		return errors.New("attempt_no must be positive")
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	numberKey := attemptNumberKey(attempt.ActionID, attempt.AttemptNo)
	if _, ok := s.attempts[attempt.ID]; ok {
		return fmt.Errorf("%w: attempt %s", ErrEntityConflict, attempt.ID)
	}
	if _, ok := s.attemptNumbers[numberKey]; ok {
		return fmt.Errorf("%w: attempt %s", ErrEntityConflict, attempt.ID)
	}
	if _, ok := s.actions[attempt.ActionID]; !ok {
		return fmt.Errorf("%w: action %s", ErrEntityNotFound, attempt.ActionID)
	}
	s.attempts[attempt.ID] = attempt
	s.attemptNumbers[numberKey] = struct{}{}
	return nil
}

func (s *MemoryEventStore) GetAttempt(ctx context.Context, id string) (Attempt, bool, error) {
	if err := ctx.Err(); err != nil {
		return Attempt{}, false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	value, ok := s.attempts[id]
	return value, ok, nil
}

func (s *MemoryEventStore) Append(ctx context.Context, laneID string, expectedLastSeq int64, appendID string, events ...ProposedEvent) (AppendReceipt, error) {
	if err := ctx.Err(); err != nil {
		return AppendReceipt{}, err
	}
	if expectedLastSeq < 0 {
		return AppendReceipt{}, errors.New("expected_last_seq must be non-negative")
	}
	if appendID == "" || len(events) == 0 {
		return AppendReceipt{}, errors.New("append requires an id and at least one event")
	}
	batch, err := clone(events)
	if err != nil {
		return AppendReceipt{}, fmt.Errorf("snapshot append batch: %w", err)
	}
	seen := make(map[string]struct{}, len(batch))
	for _, event := range batch {
		if event.LaneID != laneID {
			return AppendReceipt{}, errors.New("all events must belong to target lane")
		}
		if _, ok := seen[event.ID]; ok {
			return AppendReceipt{}, fmt.Errorf("%w: %s", ErrDuplicateEvent, event.ID)
		}
		seen[event.ID] = struct{}{}
	}
	digest, err := CanonicalAppendDigest(batch)
	if err != nil {
		return AppendReceipt{}, err
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	if previous, ok := s.appends[appendID]; ok {
		if previous.LaneID != laneID || previous.Digest != digest {
			return AppendReceipt{}, ErrIdempotencyViolation
		}
		return cloneOne(previous)
	}
	lane, ok := s.lanes[laneID]
	if !ok {
		return AppendReceipt{}, fmt.Errorf("%w: lane %s", ErrEntityNotFound, laneID)
	}
	if lane.LastSeq != expectedLastSeq {
		return AppendReceipt{}, fmt.Errorf("%w: expected %d, actual %d", ErrLaneConflict, expectedLastSeq, lane.LastSeq)
	}
	for _, event := range batch {
		if _, exists := s.events[event.ID]; exists {
			return AppendReceipt{}, fmt.Errorf("%w: %s", ErrDuplicateEvent, event.ID)
		}
		if _, exists := s.actors[event.ActorID]; !exists {
			return AppendReceipt{}, fmt.Errorf("%w: actor %s", ErrEntityNotFound, event.ActorID)
		}
		if !s.validSubject(lane, event) {
			return AppendReceipt{}, fmt.Errorf("%w: event %s", ErrSubjectMismatch, event.ID)
		}
	}
	prior := make(map[string]struct{}, len(s.events)+len(batch))
	for id := range s.events {
		prior[id] = struct{}{}
	}
	for _, event := range batch {
		if event.CausationID != "" {
			if _, exists := prior[event.CausationID]; !exists {
				return AppendReceipt{}, fmt.Errorf("%w: causation event %s", ErrEntityNotFound, event.CausationID)
			}
			if caused, exists := s.events[event.CausationID]; exists {
				causedLane := s.lanes[caused.LaneID]
				if causedLane.SessionID != lane.SessionID {
					return AppendReceipt{}, fmt.Errorf("%w: causation event %s", ErrSubjectMismatch, event.ID)
				}
			}
		}
		prior[event.ID] = struct{}{}
	}
	committedAt := now()
	stored := make([]StoredEvent, 0, len(batch))
	for index, event := range batch {
		stored = append(stored, StoredEvent{ProposedEvent: event, Seq: lane.LastSeq + int64(index) + 1, CommittedAt: committedAt})
	}
	receipt := AppendReceipt{ID: appendID, LaneID: laneID, Digest: digest, FirstSeq: stored[0].Seq, LastSeq: stored[len(stored)-1].Seq, CommittedAt: committedAt}
	for _, event := range stored {
		receipt.EventIDs = append(receipt.EventIDs, event.ID)
		s.events[event.ID] = event
	}
	lane.LastSeq = receipt.LastSeq
	s.lanes[laneID] = lane
	s.laneEvents[laneID] = append(s.laneEvents[laneID], stored...)
	s.appends[appendID] = receipt
	return cloneOne(receipt)
}

func (s *MemoryEventStore) LoadLane(ctx context.Context, laneID string, afterSeq int64) iter.Seq2[StoredEvent, error] {
	return func(yield func(StoredEvent, error) bool) {
		if afterSeq < 0 {
			yield(StoredEvent{}, errors.New("after_seq must be non-negative"))
			return
		}
		s.mu.Lock()
		_, exists := s.lanes[laneID]
		snapshot := append([]StoredEvent(nil), s.laneEvents[laneID]...)
		s.mu.Unlock()
		if !exists {
			yield(StoredEvent{}, fmt.Errorf("%w: lane %s", ErrEntityNotFound, laneID))
			return
		}
		events, err := clone(snapshot)
		if err != nil {
			yield(StoredEvent{}, err)
			return
		}
		for _, event := range events {
			if err := ctx.Err(); err != nil {
				yield(StoredEvent{}, err)
				return
			}
			if event.Seq > afterSeq && !yield(event, nil) {
				return
			}
		}
	}
}

func (s *MemoryEventStore) LoadSession(ctx context.Context, sessionID string) (SessionView, error) {
	if err := ctx.Err(); err != nil {
		return SessionView{}, err
	}
	s.mu.Lock()
	view := SessionView{SessionID: sessionID}
	laneIDs := map[string]struct{}{}
	for _, lane := range s.lanes {
		if lane.SessionID == sessionID {
			view.Lanes = append(view.Lanes, lane)
			laneIDs[lane.ID] = struct{}{}
		}
	}
	turnIDs := map[string]struct{}{}
	for _, turn := range s.turns {
		if _, ok := laneIDs[turn.LaneID]; ok {
			view.Turns = append(view.Turns, turn)
			turnIDs[turn.ID] = struct{}{}
		}
	}
	actionIDs := map[string]struct{}{}
	for _, action := range s.actions {
		if _, ok := turnIDs[action.TurnID]; ok {
			view.Actions = append(view.Actions, action)
			actionIDs[action.ID] = struct{}{}
		}
	}
	for _, attempt := range s.attempts {
		if _, ok := actionIDs[attempt.ActionID]; ok {
			view.Attempts = append(view.Attempts, attempt)
		}
	}
	for _, event := range s.events {
		if _, ok := laneIDs[event.LaneID]; ok {
			view.Events = append(view.Events, event)
		}
	}
	actorIDs := map[string]struct{}{}
	for _, event := range view.Events {
		actorIDs[event.ActorID] = struct{}{}
	}
	for _, actor := range s.actors {
		if _, ok := actorIDs[actor.ID]; ok {
			view.Actors = append(view.Actors, actor)
		}
	}
	s.mu.Unlock()
	sort.Slice(view.Events, func(i, j int) bool { return observedBefore(view.Events[i], view.Events[j]) })
	return cloneOne(view)
}

func observedBefore(left, right StoredEvent) bool {
	leftTime, leftErr := time.Parse(time.RFC3339Nano, left.CommittedAt)
	rightTime, rightErr := time.Parse(time.RFC3339Nano, right.CommittedAt)
	if leftErr == nil && rightErr == nil && !leftTime.Equal(rightTime) {
		return leftTime.Before(rightTime)
	}
	if left.CommittedAt != right.CommittedAt {
		return left.CommittedAt < right.CommittedAt
	}
	if left.LaneID != right.LaneID {
		return left.LaneID < right.LaneID
	}
	if left.Seq != right.Seq {
		return left.Seq < right.Seq
	}
	return left.ID < right.ID
}

func (s *MemoryEventStore) validSubject(lane Lane, event ProposedEvent) bool {
	switch event.SubjectKind() {
	case "session":
		return event.SubjectID == lane.SessionID
	case "run":
		return event.SubjectID == lane.RunID
	case "lane":
		return event.SubjectID == lane.ID
	case "turn":
		turn, ok := s.turns[event.SubjectID]
		return ok && turn.LaneID == lane.ID
	case "action":
		action, ok := s.actions[event.SubjectID]
		if !ok {
			return false
		}
		turn, ok := s.turns[action.TurnID]
		return ok && turn.LaneID == lane.ID
	case "attempt":
		attempt, ok := s.attempts[event.SubjectID]
		if !ok {
			return false
		}
		action, ok := s.actions[attempt.ActionID]
		if !ok {
			return false
		}
		turn, ok := s.turns[action.TurnID]
		return ok && turn.LaneID == lane.ID
	default:
		return false
	}
}

func laneNameKey(sessionID, runID, name string) string {
	return sessionID + "\x00" + runID + "\x00" + name
}
func attemptNumberKey(actionID string, attemptNo int) string {
	return fmt.Sprintf("%s\x00%d", actionID, attemptNo)
}

func clone[T any](value T) (T, error) {
	var result T
	encoded, err := json.Marshal(value)
	if err != nil {
		return result, err
	}
	if err := json.Unmarshal(encoded, &result); err != nil {
		return result, err
	}
	return result, nil
}

func cloneOne[T any](value T) (T, error) { return clone(value) }

File: typescript/packages/core/src/memory-store.ts
import { canonicalAppendDigest, canonicalize } from "./canonical.js";
import type { CheckpointStore, EventStore } from "./store.js";
import type {
  Action, Actor, AppendReceipt, Attempt, Checkpoint, Lane, ProposedCheckpoint, ProposedEvent,
  RunView, SessionView, StoredEvent, Turn,
} from "./types.js";
import { selectRun } from "./types.js";

export class LaneConflict extends Error {}
export class CheckpointConflict extends Error {}
export class IdempotencyViolation extends Error {}
export class CheckpointIdempotencyViolation extends Error {}
export class DuplicateEvent extends Error {}
export class EntityConflict extends Error {}
export class EntityNotFound extends Error {}
export class SubjectMismatch extends Error {}

export class MemoryEventStore implements EventStore, CheckpointStore {
  readonly #actors = new Map<string, Actor>();
  readonly #actorKeys = new Map<string, string>();
  readonly #lanes = new Map<string, Lane>();
  readonly #laneNames = new Map<string, string>();
  readonly #turns = new Map<string, Turn>();
  readonly #actions = new Map<string, Action>();
  readonly #attempts = new Map<string, Attempt>();
  readonly #attemptNumbers = new Set<string>();
  readonly #events = new Map<string, StoredEvent>();
  readonly #laneEvents = new Map<string, StoredEvent[]>();
  readonly #appends = new Map<string, AppendReceipt>();
  readonly #checkpoints = new Map<string, Checkpoint>();
  readonly #latestCheckpoints = new Map<string, string>();

  async createActor(actor: Actor): Promise<void> {
    if (actor.key !== undefined && this.#actorKeys.has(actor.key)) throw new EntityConflict(`actor key ${actor.key}`);
    this.#create(this.#actors, "actor", actor.id, actor);
    if (actor.key !== undefined) this.#actorKeys.set(actor.key, actor.id);
  }

  async getActor(id: string): Promise<Actor | undefined> {
    return cloneOptional(this.#actors.get(id));
  }

  async getActorByKey(key: string): Promise<Actor | undefined> {
    const id = this.#actorKeys.get(key);
    return id === undefined ? undefined : cloneOptional(this.#actors.get(id));
  }

  async ensureActor(actor: Actor): Promise<Actor> {
    const stored = actor.key === undefined
      ? this.#actors.get(actor.id)
      : this.#actors.get(this.#actorKeys.get(actor.key) ?? "");
    if (stored !== undefined) {
      requireSameActor(stored, actor);
      return structuredClone(stored);
    }
    await this.createActor(actor);
    return structuredClone(actor);
  }

  async createLane(lane: Lane): Promise<void> {
    if (lane.last_seq !== 0) throw new TypeError("new lane must have last_seq 0");
    if (this.#lanes.has(lane.id) || this.#laneNames.has(laneNameKey(lane.session_id, lane.run_id, lane.name))) {
      throw new EntityConflict(`lane ${lane.id}`);
    }
    if (lane.parent_lane_id !== undefined) {
      const parent = this.#lanes.get(lane.parent_lane_id);
      if (!parent) throw new EntityNotFound(`parent lane ${lane.parent_lane_id}`);
      if (parent.session_id !== lane.session_id) throw new TypeError("parent lane must belong to same session");
    }
    this.#lanes.set(lane.id, structuredClone(lane));
    this.#laneNames.set(laneNameKey(lane.session_id, lane.run_id, lane.name), lane.id);
  }

  async getLane(id: string): Promise<Lane | undefined> {
    return cloneOptional(this.#lanes.get(id));
  }

  async findLane(sessionId: string, runId: string, name: string): Promise<Lane | undefined> {
    const id = this.#laneNames.get(laneNameKey(sessionId, runId, name));
    return id === undefined ? undefined : structuredClone(this.#lanes.get(id)!);
  }

  async createTurn(turn: Turn): Promise<void> {
    if (!this.#lanes.has(turn.lane_id)) throw new EntityNotFound(`lane ${turn.lane_id}`);
    this.#create(this.#turns, "turn", turn.id, turn);
  }

  async getTurn(id: string): Promise<Turn | undefined> {
    return cloneOptional(this.#turns.get(id));
  }

  async createAction(action: Action): Promise<void> {
    if (!this.#turns.has(action.turn_id)) throw new EntityNotFound(`turn ${action.turn_id}`);
    if (action.parent_action_id !== undefined) {
      const parent = this.#actions.get(action.parent_action_id);
      if (!parent) throw new EntityNotFound(`parent action ${action.parent_action_id}`);
      if (parent.turn_id !== action.turn_id) throw new TypeError("parent action must belong to same turn");
    }
    this.#create(this.#actions, "action", action.id, action);
  }

  async getAction(id: string): Promise<Action | undefined> {
    return cloneOptional(this.#actions.get(id));
  }

  async createAttempt(attempt: Attempt): Promise<void> {
    if (attempt.attempt_no < 1) throw new TypeError("attempt_no must be positive");
    if (!this.#actions.has(attempt.action_id)) throw new EntityNotFound(`action ${attempt.action_id}`);
    const numberKey = `${attempt.action_id}\u0000${attempt.attempt_no}`;
    if (this.#attemptNumbers.has(numberKey)) throw new EntityConflict(`attempt ${attempt.id}`);
    this.#create(this.#attempts, "attempt", attempt.id, attempt);
    this.#attemptNumbers.add(numberKey);
  }

  async getAttempt(id: string): Promise<Attempt | undefined> {
    return cloneOptional(this.#attempts.get(id));
  }

  async saveCheckpoint(expectedRevision: number, proposed: ProposedCheckpoint): Promise<Checkpoint> {
    if (!Number.isSafeInteger(expectedRevision) || expectedRevision < 0) {
      throw new TypeError("expectedRevision must be a non-negative safe integer");
    }
    validateCheckpoint(proposed);
    const checkpoint = structuredClone(proposed);
    const previous = this.#checkpoints.get(checkpoint.id);
    if (previous) {
      if (!sameCheckpoint(previous, checkpoint)) throw new CheckpointIdempotencyViolation(checkpoint.id);
      return structuredClone(previous);
    }
    if (!this.#actors.has(checkpoint.actor_id)) throw new EntityNotFound(`actor ${checkpoint.actor_id}`);
    const latestId = this.#latestCheckpoints.get(checkpoint.key);
    const actualRevision = latestId === undefined ? 0 : this.#checkpoints.get(latestId)!.revision;
    if (actualRevision !== expectedRevision) {
      throw new CheckpointConflict(`expected ${expectedRevision}, actual ${actualRevision}`);
    }
    if (checkpoint.anchor !== undefined) {
      const event = this.#events.get(checkpoint.anchor.last_applied_event_id);
      if (event === undefined
        || event.lane_id !== checkpoint.anchor.lane_id
        || event.seq !== checkpoint.anchor.last_applied_seq) {
        throw new TypeError("checkpoint anchor must identify an existing lane event");
      }
    }
    const stored: Checkpoint = {
      ...checkpoint,
      revision: actualRevision + 1,
      created_at: new Date().toISOString(),
    };
    this.#checkpoints.set(stored.id, stored);
    this.#latestCheckpoints.set(stored.key, stored.id);
    return structuredClone(stored);
  }

  async getCheckpoint(id: string): Promise<Checkpoint | undefined> {
    return cloneOptional(this.#checkpoints.get(id));
  }

  async loadLatestCheckpoint(key: string): Promise<Checkpoint | undefined> {
    const id = this.#latestCheckpoints.get(key);
    return id === undefined ? undefined : structuredClone(this.#checkpoints.get(id)!);
  }

  async append(
    laneId: string,
    expectedLastSeq: number,
    appendId: string,
    events: readonly ProposedEvent[],
  ): Promise<AppendReceipt> {
    if (expectedLastSeq < 0) throw new TypeError("expectedLastSeq must be non-negative");
    if (appendId === "" || events.length === 0) throw new TypeError("append requires an id and at least one event");
    const batch = structuredClone(events) as ProposedEvent[];
    const digest = canonicalAppendDigest(batch);
    const previous = this.#appends.get(appendId);
    if (previous) {
      if (previous.lane_id !== laneId || previous.digest !== digest) throw new IdempotencyViolation(appendId);
      return structuredClone(previous);
    }
    const lane = this.#lanes.get(laneId);
    if (!lane) throw new EntityNotFound(`lane ${laneId}`);
    if (lane.last_seq !== expectedLastSeq) throw new LaneConflict(laneId);
    const prior = new Set(this.#events.keys());
    const batchIds = new Set<string>();
    for (const event of batch) {
      if (event.lane_id !== laneId) throw new TypeError("all events must belong to target lane");
      if (batchIds.has(event.id) || this.#events.has(event.id)) throw new DuplicateEvent(event.id);
      batchIds.add(event.id);
      if (!this.#actors.has(event.actor_id)) throw new EntityNotFound(`actor ${event.actor_id}`);
      if (!this.#validSubject(lane, event)) throw new SubjectMismatch(event.id);
      if (event.causation_id !== undefined) {
        if (!prior.has(event.causation_id)) throw new EntityNotFound(`causation event ${event.causation_id}`);
        const caused = this.#events.get(event.causation_id);
        if (caused && this.#lanes.get(caused.lane_id)?.session_id !== lane.session_id) {
          throw new SubjectMismatch(event.id);
        }
      }
      prior.add(event.id);
    }

    const committedAt = new Date().toISOString();
    const stored = batch.map((event, index): StoredEvent => ({
      ...event, seq: lane.last_seq + index + 1, committed_at: committedAt,
    }));
    const receipt: AppendReceipt = {
      id: appendId, lane_id: laneId, digest, first_seq: stored[0]!.seq,
      last_seq: stored.at(-1)!.seq, event_ids: stored.map((event) => event.id), committed_at: committedAt,
    };
    for (const event of stored) this.#events.set(event.id, event);
    lane.last_seq = receipt.last_seq;
    this.#laneEvents.set(laneId, [...(this.#laneEvents.get(laneId) ?? []), ...stored]);
    this.#appends.set(appendId, receipt);
    return structuredClone(receipt);
  }

  async *loadLane(laneId: string, afterSeq = 0): AsyncIterable<StoredEvent> {
    if (afterSeq < 0) throw new TypeError("afterSeq must be non-negative");
    if (!this.#lanes.has(laneId)) throw new EntityNotFound(`lane ${laneId}`);
    for (const event of this.#laneEvents.get(laneId) ?? []) {
      if (event.seq > afterSeq) yield structuredClone(event);
    }
  }

  async loadSession(sessionId: string): Promise<SessionView> {
    const lanes = [...this.#lanes.values()].filter((lane) => lane.session_id === sessionId);
    const laneIds = new Set(lanes.map((lane) => lane.id));
    const turns = [...this.#turns.values()].filter((turn) => laneIds.has(turn.lane_id));
    const turnIds = new Set(turns.map((turn) => turn.id));
    const actions = [...this.#actions.values()].filter((action) => turnIds.has(action.turn_id));
    const actionIds = new Set(actions.map((action) => action.id));
    const attempts = [...this.#attempts.values()].filter((attempt) => actionIds.has(attempt.action_id));
    const events = [...this.#events.values()]
      .filter((event) => laneIds.has(event.lane_id))
      .sort((left, right) => left.committed_at.localeCompare(right.committed_at)
        || left.lane_id.localeCompare(right.lane_id)
        || left.seq - right.seq
        || left.id.localeCompare(right.id));
    const actorIds = new Set(events.map((event) => event.actor_id));
    const actors = [...this.#actors.values()].filter((actor) => actorIds.has(actor.id));
    return structuredClone({ session_id: sessionId, actors, lanes, turns, actions, attempts, events });
  }

  async loadRun(sessionId: string, runId: string): Promise<RunView> {
    return selectRun(await this.loadSession(sessionId), runId);
  }

  #create<T>(target: Map<string, T>, kind: string, id: string, value: T): void {
    if (target.has(id)) throw new EntityConflict(`${kind} ${id}`);
    target.set(id, structuredClone(value));
  }

  #validSubject(lane: Lane, event: ProposedEvent): boolean {
    const kind = event.event_type.split(".", 1)[0];
    if (kind === "session") return event.subject_id === lane.session_id;
    if (kind === "run") return event.subject_id === lane.run_id;
    if (kind === "lane") return event.subject_id === lane.id;
    if (kind === "turn") return this.#turns.get(event.subject_id)?.lane_id === lane.id;
    if (kind === "action") {
      const action = this.#actions.get(event.subject_id);
      return action !== undefined && this.#turns.get(action.turn_id)?.lane_id === lane.id;
    }
    if (kind === "attempt") {
      const attempt = this.#attempts.get(event.subject_id);
      const action = attempt === undefined ? undefined : this.#actions.get(attempt.action_id);
      return action !== undefined && this.#turns.get(action.turn_id)?.lane_id === lane.id;
    }
    return false;
  }
}

function laneNameKey(sessionId: string, runId: string, name: string): string {
  return `${sessionId}\u0000${runId}\u0000${name}`;
}

function cloneOptional<T>(value: T | undefined): T | undefined {
  return value === undefined ? undefined : structuredClone(value);
}

function requireSameActor(stored: Actor, proposed: Actor): void {
  if (stored.key !== proposed.key || stored.type !== proposed.type || stored.framework !== proposed.framework) {
    throw new EntityConflict(`actor key ${proposed.key ?? proposed.id}`);
  }
}

function validateCheckpoint(value: ProposedCheckpoint): void {
  if (value.schema_version !== "1.0" || value.id === "" || value.key === ""
    || value.actor_id === "" || value.format === "") {
    throw new TypeError("checkpoint requires schema version, id, key, actor, and format");
  }
  if (("state" in value) === ("artifact_ref" in value)) {
    throw new TypeError("exactly one of state and artifact_ref must be set");
  }
  if (value.anchor !== undefined && (!Number.isSafeInteger(value.anchor.last_applied_seq)
    || value.anchor.last_applied_seq < 1 || value.anchor.lane_id === ""
    || value.anchor.last_applied_event_id === "")) {
    throw new TypeError("checkpoint anchor requires lane, positive seq, and event");
  }
}

function sameCheckpoint(stored: Checkpoint, proposed: ProposedCheckpoint): boolean {
  const { revision: _revision, created_at: _createdAt, ...existing } = stored;
  return canonicalize(existing) === canonicalize(proposed);
}
